File: src/main/java/org/codetome/hexameter/core/internal/impl/HexagonImpl.java
package org.codetome.hexameter.core.internal.impl;

import org.codetome.hexameter.core.api.*;
import org.codetome.hexameter.core.api.contract.HexagonDataStorage;
import org.codetome.hexameter.core.api.contract.SatelliteData;
import org.codetome.hexameter.core.backport.Optional;
import org.codetome.hexameter.core.internal.GridData;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static java.lang.Math.cos;
import static java.lang.Math.sin;

/**
 * Default implementation of the {@link Hexagon} interface.
 */
public final class HexagonImpl<T extends SatelliteData> implements Hexagon<T> {
    private final CubeCoordinate coordinate;
    private final transient List<Double> vertices;
    private final transient List<Point> points;
    private final transient Rectangle externalBoundingBox;
    private final transient Rectangle internalBoundingBox;
    private final transient GridData sharedData;
    private final transient HexagonDataStorage<T> hexagonDataStorage;

    /**
     * Creates a new {@link Hexagon} object from shared data and a coordinate.
     *
     * @param gridData           grid data
     * @param coordinate         coordinate
     * @param hexagonDataStorage data map
     **/
    HexagonImpl(final GridData gridData, final CubeCoordinate coordinate, final HexagonDataStorage<T> hexagonDataStorage) {
        this.sharedData = gridData;
        this.coordinate = coordinate;
        this.hexagonDataStorage = hexagonDataStorage;

        this.points = calculatePoints();
        final int x1 = (int) points.get(3).getCoordinateX();
        final int y1 = (int) points.get(2).getCoordinateY();
        final int x2 = (int) points.get(0).getCoordinateX();
        final int y2 = (int) points.get(5).getCoordinateY();

        externalBoundingBox = new Rectangle(x1, y1, x2 - x1, y2 - y1);
        internalBoundingBox = new Rectangle((int) (getCenterX() - (1.25 * sharedData.getRadius() / 2)),
                (int) (getCenterY() - (1.25 * sharedData.getRadius() / 2)),
                (int) (1.25f * sharedData.getRadius()),
                (int) (1.25f * sharedData.getRadius()));

        this.vertices = new ArrayList<>();
        for(Point point : points) {
            vertices.add(point.getCoordinateX());
            vertices.add(point.getCoordinateY());
        }
    }

    private List<Point> calculatePoints() {
        final List<Point> points = new ArrayList<>();
        for (int i = 0; i < 6; i++) {
            final double angle = 2 * Math.PI / 6 * (i + sharedData.getOrientation().getCoordinateOffset());
            final double x = getCenterX() + sharedData.getRadius() * cos(angle);
            final double y = getCenterY() + sharedData.getRadius() * sin(angle);
            points.add(Point.fromPosition(x, y));
        }
        return points;
    }

    @Override
    public String getId() {
        return coordinate.toAxialKey();
    }

    @Override
    public List<Point> getPoints() {
        return points;
    }

    @Override
    public List<Double> getVertices() {
        return vertices;
    }

    @Override
    public Rectangle getExternalBoundingBox() {
        return externalBoundingBox;
    }

    @Override
    public Rectangle getInternalBoundingBox() {
        return internalBoundingBox;
    }

    @Override
    public CubeCoordinate getCubeCoordinate() {
        return coordinate;
    }

    @Override
    public int getGridX() {
        return coordinate.getGridX();
    }

    @Override
    public int getGridY() {
        return coordinate.getGridY();
    }

    @Override
    public int getGridZ() {
        return coordinate.getGridZ();
    }

    @Override
    public double getCenterX() {
        if (HexagonOrientation.FLAT_TOP.equals(sharedData.getOrientation())) {
            return coordinate.getGridX() * sharedData.getHexagonWidth() + sharedData.getRadius();
        } else {
            return coordinate.getGridX() * sharedData.getHexagonWidth() + coordinate.getGridZ()
                    * sharedData.getHexagonWidth() / 2 + sharedData.getHexagonWidth() / 2;
        }
    }

    @Override
    public double getCenterY() {
        if (HexagonOrientation.FLAT_TOP.equals(sharedData.getOrientation())) {
            return coordinate.getGridZ() * sharedData.getHexagonHeight() + coordinate.getGridX()
                    * sharedData.getHexagonHeight() / 2 + sharedData.getHexagonHeight() / 2;
        } else {
            return coordinate.getGridZ() * sharedData.getHexagonHeight() + sharedData.getRadius();
        }
    }

    @Override
    public Optional<T> getSatelliteData() {
        return hexagonDataStorage.getSatelliteDataBy(getCubeCoordinate());
    }

    @Override
    public void setSatelliteData(final T satelliteData) {
        this.hexagonDataStorage.addCoordinate(getCubeCoordinate(), satelliteData);
    }

    @Override
    public void clearSatelliteData() {
        this.hexagonDataStorage.clearDataFor(getCubeCoordinate());
    }

    @Override
    public int hashCode() {
        return Objects.hash(coordinate);
    }

    @Override
    public boolean equals(final Object object) {
        if (this == object) {
            return true;
        }
        if (object == null || getClass() != object.getClass()) {
            return false;
        }
        final HexagonImpl hexagon = (HexagonImpl) object;
        return Objects.equals(coordinate, hexagon.coordinate);
    }
}

File: mixite.core/src/commonMain/kotlin/org/hexworks/mixite/core/internal/impl/HexagonImpl.kt
package org.hexworks.mixite.core.internal.impl

import org.hexworks.mixite.core.api.*
import org.hexworks.mixite.core.api.contract.HexagonDataStorage
import org.hexworks.mixite.core.api.contract.SatelliteData
import org.hexworks.mixite.core.internal.GridData
import org.hexworks.mixite.core.vendor.Maybe
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * Default implementation of the [Hexagon] interface.
 */
class HexagonImpl<T : SatelliteData> internal constructor(
        private val sharedData: GridData,
        override val cubeCoordinate: CubeCoordinate,
        private val hexagonDataStorage: HexagonDataStorage<T>) : Hexagon<T> {

    override val vertices: MutableList<Double>
    override val points: List<Point>
    override val externalBoundingBox: Rectangle
    override val internalBoundingBox: Rectangle

    override val id: String
        get() = cubeCoordinate.toAxialKey()

    override val gridX: Int
        get() = cubeCoordinate.gridX

    override val gridY: Int
        get() = cubeCoordinate.gridY

    override val gridZ: Int
        get() = cubeCoordinate.gridZ

    override val centerX: Double
        get() = if (HexagonOrientation.FLAT_TOP.equals(sharedData.orientation)) {
            cubeCoordinate.gridX * sharedData.hexagonWidth + sharedData.radius
        } else {
            cubeCoordinate.gridX * sharedData.hexagonWidth + cubeCoordinate.gridZ * sharedData.hexagonWidth / 2 + sharedData.hexagonWidth / 2
        }

    override val centerY: Double
        get() = if (HexagonOrientation.FLAT_TOP.equals(sharedData.orientation)) {
            cubeCoordinate.gridZ * sharedData.hexagonHeight + cubeCoordinate.gridX * sharedData.hexagonHeight / 2 + sharedData.hexagonHeight / 2
        } else {
            cubeCoordinate.gridZ * sharedData.hexagonHeight + sharedData.radius
        }

    override val satelliteData: Maybe<T>
        get() = hexagonDataStorage.getSatelliteDataBy(cubeCoordinate)

    init {

        this.points = calculatePoints()
        val x1 = points[3].coordinateX
        val y1 = points[2].coordinateY
        val x2 = points[0].coordinateX
        val y2 = points[5].coordinateY

        externalBoundingBox = Rectangle(x1, y1, x2 - x1, y2 - y1)
        internalBoundingBox = Rectangle((centerX - 1.25 * sharedData.radius / 2),
                (centerY - 1.25 * sharedData.radius / 2),
                (1.25f * sharedData.radius),
                (1.25f * sharedData.radius))

        this.vertices = ArrayList()
        for (point in points) {
            vertices.add(point.coordinateX)
            vertices.add(point.coordinateY)
        }
    }

    private fun calculatePoints(): List<Point> {
        val points = ArrayList<Point>()
        for (i in 0..5) {
            val angle = 2 * PI / 6 * (i + sharedData.orientation.coordinateOffset)
            val x = centerX + sharedData.radius * cos(angle)
            val y = centerY + sharedData.radius * sin(angle)
            points.add(Point.fromPosition(x, y))
        }
        return points
    }

    override fun setSatelliteData(data: T) {
        this.hexagonDataStorage.addCoordinate(cubeCoordinate, data)
    }

    override fun clearSatelliteData() {
        this.hexagonDataStorage.clearDataFor(cubeCoordinate)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false

        other as HexagonImpl<*>

        if (cubeCoordinate != other.cubeCoordinate) return false

        return true
    }

    override fun hashCode(): Int {
        return cubeCoordinate.hashCode()
    }


}
